lib/AST/DiagnosticEngine.cpp

```diff
@@ -311,6 +313,38 @@ void Diagnostic::addChildNote(Diagnostic &&D) {
   ChildNotes.push_back(std::move(D));
 }
 
+YAMLDiagnosticProvider::YAMLDiagnosticProvider(std::string locale,
+                                               std::string path) {
+  llvm::SmallString<128> DiagnosticsFilePath(path);
+  llvm::sys::path::append(DiagnosticsFilePath, locale);
+  llvm::sys::path::replace_extension(DiagnosticsFilePath, ".yaml");
+  auto FileBufOrErr = llvm::MemoryBuffer::getFileOrSTDIN(DiagnosticsFilePath);
+  if (!FileBufOrErr)
+    llvm_unreachable("Failed to read yaml file");
+
+  std::vector<DiagnosticNode> Diags;
+  llvm::MemoryBuffer *document = FileBufOrErr->get();
+  llvm::yaml::Input yin(document->getBuffer());
+  yin >> Diags;
```

Comment:
Is it possible to read diagnostics from YAML file one-by-one instead of in-bulk like this using `>>` operator? If could be read iteratively that would make a separate temporary vector used here redundant.

Reply:
I don't think we can do that, at least as far as I understand. 

In the [`LLVM::YAML I/O` documentation](https://llvm.org/docs/YamlIO.html) reading the `YAML` file in bulk is the only way mentioned.  

I also took a look at the source code of `LLVM->YAMLTraits.h` and found out that the way the `>>` operator work is by using two methods called `setCurrentDocument` and `nextDocument`. And according to this comment, they might change the access modifier of both of these methods to be private. Therefore, I don't think there is a better way than reading the `YAML` file in bulk, I'd love to hear your thoughts?

**Links**
[Here's how `<<` operator is implemented](https://github.com/llvm/llvm-project/blob/master/llvm/include/llvm/Support/YAMLTraits.h#L1715).
[Here are the methods for it.](https://github.com/llvm/llvm-project/blob/master/llvm/include/llvm/Support/YAMLTraits.h#L1517)

Reply:
Ok so how about a custom `yaml::Input` which would know how to output everything in correct order?

Reply:
So it could do `yin >> diagnostics;` directly.

Reply:
This is a good idea! I’ll look into it tomorrow. 

But just to be on the same page.. I will override the `>>` operator get diagnostics in order. Right?

Reply:
Yes, you need to add a new class which is going to inherit from `yaml::Input` and implement a `>>` operator.